File: 2021/Day2.hs
#! /usr/bin/env -S stack runhaskell
import Data.Char (toUpper)

data Command = Forward Int | Up Int | Down Int deriving (Show, Read)

part1 :: [Command] -> Int
part1 = go 0 0
    where
        go d p [] = d * p
        go d p ((Forward x):xs) = go d (p+x) xs
        go d p ((Up x):xs) = go (d-x) p xs
        go d p ((Down x):xs) = go (d+x) p xs

part2 :: [Command] -> Int
part2 = go 0 0
    where
        go d p a [] = d * p
        go d p a ((Forward x):xs) = go (d+a*x) (p+x) a xs
        go d p a ((Up x):xs) = go d p (a-x) xs
        go d p a ((Down x):xs) = go d p (a+x) xs

main :: IO()
main = do
    f <- readFile("2.in")
    let input = map read . map (\(x:xs) -> (toUpper x) : xs) $ lines f :: [Command]
    putStr "Solution for part 1: "
    putStrLn . show $ part1 input
    putStr "Solution for part 2: "
    putStrLn . show $ part2 input

-- Solution part 1: 1804520
-- Solution part 2: 1971095320

File: 2021/Day2.py
#! /usr/bin/env python
# pylint: disable=unused-wildcard-import
from util import *

def p1(inp):
    d = 0
    p = 0
    for xs in inp:
        cmd, x = xs
        
        if cmd == 'forward':
            p += x
        if cmd == 'down':
            d += x
        if cmd == 'up':
            d -= x
    return d * p

def p2(inp):
    d = 0
    p = 0
    a = 0
    for xs in inp:
        cmd, x = xs
        
        if cmd == 'forward':
            p += x
            d += a*x
        if cmd == 'down':
            a += x
        if cmd == 'up':
            a -= x
    return d * p

def main():
    inp = structured(data(), (str, int))

    print(f"Solution for part 1:\n{p1(inp)}")
    print(f"Solution for part 2:\n{p2(inp)}")

if __name__ == "__main__":
    main()

# Solution part 1: 1804520
# Solution part 2: 1971095320
